File: Bindings/Python/examples/Moco/exampleSquatToStand/mocoPlotTrajectory.py
import opensim as osim
import matplotlib.pyplot as plt
import numpy as np

# Plot a MocoTrajectory. Optionally, specify a second trajectory and names
# for the trajectories.
def mocoPlotTrajectory(*args):
    args_list = list(args)

    if isinstance(args_list[0], str):
        trajA = osim.MocoTrajectory(args_list[0])
    
    if len(args) > 1 and isinstance(args_list[1], str):
        trajB = osim.MocoTrajectory(args_list[1])

    labelA = args_list[2] if len(args) == 4 else 'A'
    labelB = args_list[3] if len(args) == 4 else 'B'
    
    ## States.
    stateNames = list(trajA.getStateNames())
    numStates = len(stateNames)
    dim = np.sqrt(numStates)
    if dim == np.ceil(dim):
        numRows = int(dim)
        numCols = int(dim)
    else:
        numCols = int(np.min([numStates, 4]))
        numRows = int(np.floor(numStates / 4))
        if not numStates % 4 == 0:
            numRows += 1
    
    fig = plt.figure()
    for i in np.arange(numStates):
        ax = fig.add_subplot(numRows, numCols, int(i+1))
        ax.plot(trajA.getTimeMat(), 
             trajA.getStateMat(stateNames[i]), color='red',
             linewidth=3, label=labelA)

        if len(args) > 1:
            ax.plot(trajB.getTimeMat(), 
                 trajB.getStateMat(stateNames[i]), 
                 color='blue', linestyle='--',
                 linewidth=2.5, label=labelB)
        
        stateName = stateNames[i]
        ax.set_title(stateName[11:])
        ax.set_xlabel('time (s)')
        ax.set_xlim(0, 1)
        if 'value' in stateName:
            ax.set_ylabel('position (rad)')
        elif 'speed' in stateName:
            ax.set_ylabel('speed (rad/s)')
        elif 'activation' in stateName:
            ax.set_ylabel('activation (-)')
            ax.set_ylim(0, 1)
        
        if i == 0 and len(args) > 1:
            ax.legend(loc='best')

    plt.show()

    ## Controls.
    controlNames = list(trajA.getControlNames())
    numControls = len(controlNames)
    dim = np.sqrt(numControls)
    if dim == np.ceil(dim):
        numRows = int(dim)
        numCols = int(dim)
    else:
        numCols = int(np.min([numControls, 4]))
        numRows = int(np.floor(numControls / 4.0))
        if not numControls % 4 == 0:
            numRows += 1

    fig = plt.figure()
    for i in np.arange(numControls):
        ax = fig.add_subplot(numRows, numCols, int(i+1))
        yA = trajA.getControlMat(controlNames[i])
        ax.plot(trajA.getTimeMat(), yA, color='red', 
            linewidth=3, label=labelA)
        if len(args) > 1:
            yB = trajB.getControlMat(controlNames[i])
            ax.plot(trajB.getTimeMat(), yB, color='blue',
                linestyle='--', linewidth=2.5, 
                label=labelB)
       
        ax.set_title(controlNames[i])
        ax.set_xlabel('time (s)')
        ax.set_ylabel('value')
        ax.set_xlim(0, 1)
        if np.max(yA) <= 1 and np.min(yA) >= 0:
            fixYLim = True
            if len(args) > 1 and (np.max(yB) > 1 or 
                                  np.min(yB) < 0):
                fixYLim = False
            
            if fixYLim:
                ax.set_ylim(0, 1)

        if i == 0 and len(args) > 1:
            ax.legend(loc='best')

    fig.tight_layout()
    plt.show()
    plt.close()

File: Bindings/Java/Matlab/examples/Moco/exampleSquatToStand/mocoPlotTrajectory.m
function mocoPlotTrajectory(trajA, trajB, nameA, nameB)
% Plot a MocoTrajectory. Optionally, specify a second trajectory and names
% for the trajectories.
% For a generic version of this function, see the utility
% osimMocoTrajectoryReport.m.

import org.opensim.modeling.*;

if ischar(trajA)
    trajA = MocoTrajectory(trajA);
end
if nargin > 1 && ischar(trajB)
    trajB = MocoTrajectory(trajB);
end
%% States.
figure;
stateNames = trajA.getStateNames();
numStates = stateNames.size();
dim = sqrt(numStates);
if dim == ceil(dim)
    numRows = dim;
    numCols = dim;
else
    numCols = min(numStates, 4);
    numRows = floor(numStates / 4);
    if mod(numStates, 4) ~= 0
        numRows = numRows + 1;
    end
end
for i = 0:numStates-1
    subplot(numRows, numCols, i+1);
    plot(trajA.getTimeMat(), ...
         trajA.getStateMat(stateNames.get(i)), '-r', ...
         'linewidth', 3);
    if nargin > 1
        hold on
        plot(trajB.getTimeMat(), ...
             trajB.getStateMat(stateNames.get(i)), '--b', ...
             'linewidth', 2.5);
        hold off
    end
    
    stateName = stateNames.get(i);
    title(stateName(11:end), 'Interpreter', 'none')
    xlabel('time (s)')
    if contains(stateName, 'value')
        ylabel('position (rad)')
    elseif contains(stateName, 'speed')
        ylabel('speed (rad/s)')
    elseif contains(stateName, 'activation')
        ylabel('activation (-)')
        ylim([0, 1])
    end
    if i == 0 && nargin > 1
        if nargin == 4
            legend(nameA, nameB);
        else
            legend('A', 'B');
        end
    end
end

%% Controls.
figure;
controlNames = trajA.getControlNames();
numControls = controlNames.size();
dim = sqrt(numControls);
if dim == ceil(dim)
    numRows = dim;
    numCols = dim;
else
    numCols = min(numControls, 4);
    numRows = floor(numControls / 4);
    if mod(numControls, 4) ~= 0
        numRows = numRows + 1;
    end
end
for i = 0:numControls-1
    subplot(numRows, numCols, i+1);
    yA = trajA.getControlMat(controlNames.get(i));
    plot(trajA.getTimeMat(), yA, '-r', 'linewidth', 3);
    if nargin > 1
        hold on
        yB = trajB.getControlMat(controlNames.get(i));
        plot(trajB.getTimeMat(), yB, '--b', 'linewidth', 2.5);
        hold off
    end
    title(controlNames.get(i), 'Interpreter', 'none')
    xlabel('time (s)')
    ylabel('value')
    if max(yA) <= 1 && min(yA) >= 0
        fixYLim = true;
        if nargin > 1 && (max(yB) > 1 || min(yB) < 0)
            fixYLim = false;
        end
        if fixYLim
            ylim([0, 1]);
        end
    end
    if i == 0 && nargin > 1
        if nargin == 4
            legend(nameA, nameB);
        else
            legend('A', 'B');
        end
    end
end
